tools/system_libs.py

```diff
@@ -78,161 +78,299 @@ def create_lib(libname, inputs):
     raise Exception('unknown suffix ' + libname)
 
 
-def calculate(temp_files, in_temp, stdout_, stderr_, forced=[]):
-  global stdout, stderr
-  stdout = stdout_
-  stderr = stderr_
+def read_symbols(path):
+  with open(path) as f:
+    content = f.read()
+
+    # Require that Windows newlines should not be present in a symbols file, if running on Linux or macOS
+    # This kind of mismatch can occur if one copies a zip file of Emscripten cloned on Windows over to
+    # a Linux or macOS system. It will result in Emscripten linker getting confused on stray \r characters,
+    # and be unable to link any library symbols properly. We could harden against this by .strip()ping the
+    # opened files, but it is possible that the mismatching line endings can cause random problems elsewhere
+    # in the toolchain, hence abort execution if so.
+    if os.name != 'nt' and '\r\n' in content:
+      raise Exception('Windows newlines \\r\\n detected in symbols file "' + path + '"! This could happen for example when copying Emscripten checkout from Windows to Linux or macOS. Please use Unix line endings on checkouts of Emscripten on Linux and macOS!')
+
+    return shared.Building.parse_symbols(content).defs
+
+
+def get_wasm_libc_rt_files():
+  # Static linking is tricky with LLVM, since e.g. memset might not be used
+  # from libc, but be used as an intrinsic, and codegen will generate a libc
+  # call from that intrinsic *after* static linking would have thought it is
+  # all in there. In asm.js this is not an issue as we do JS linking anyhow,
+  # and have asm.js-optimized versions of all the LLVM intrinsics. But for
+  # wasm, we need a better solution. For now, make another archive that gets
+  # included at the same time as compiler-rt.
+  # Note that this also includes things that may be depended on by those
+  # functions - fmin uses signbit, for example, so signbit must be here (so if
+  # fmin is added by codegen, it will have all it needs).
+  math_files = files_in_path(
+    path_components=['system', 'lib', 'libc', 'musl', 'src', 'math'],
+    filenames=[
+      'fmin.c', 'fminf.c', 'fminl.c',
+      'fmax.c', 'fmaxf.c', 'fmaxl.c',
+      'fmod.c', 'fmodf.c', 'fmodl.c',
+      'log2.c', 'log2f.c', 'log10.c', 'log10f.c',
+      'exp2.c', 'exp2f.c', 'exp10.c', 'exp10f.c',
+      'scalbn.c', '__fpclassifyl.c',
+      '__signbitl.c', '__signbitf.c', '__signbit.c'
+    ])
+  string_files = files_in_path(
+    path_components=['system', 'lib', 'libc', 'musl', 'src', 'string'],
+    filenames=['memset.c', 'memmove.c'])
+  other_files = files_in_path(
+    path_components=['system', 'lib', 'libc'],
+    filenames=['emscripten_memcpy.c'])
+  return math_files + string_files + other_files
+
+
+class Library(object):
+  # A name that is not None means a concrete library and not an abstract one.
+  name = None
+  depends = []
+  symbols = set()
+  js_depends = []
+
```

Comment:
would be good to add some comments to explain what these fields are

Reply:
Done.